datafusion/sqllogictest/test_files/tpch/plans/q6.slt.part

```diff
@@ -31,13 +31,13 @@ logical_plan
 01)Projection: sum(lineitem.l_extendedprice * lineitem.l_discount) AS revenue
 02)--Aggregate: groupBy=[[]], aggr=[[sum(lineitem.l_extendedprice * lineitem.l_discount)]]
 03)----Projection: lineitem.l_extendedprice, lineitem.l_discount
-04)------Filter: lineitem.l_shipdate >= Date32("1994-01-01") AND lineitem.l_shipdate < Date32("1995-01-01") AND lineitem.l_discount >= Decimal128(Some(5),15,2) AND lineitem.l_discount <= Decimal128(Some(7),15,2) AND lineitem.l_quantity < Decimal128(Some(2400),15,2)
-05)--------TableScan: lineitem projection=[l_quantity, l_extendedprice, l_discount, l_shipdate], partial_filters=[lineitem.l_shipdate >= Date32("1994-01-01"), lineitem.l_shipdate < Date32("1995-01-01"), lineitem.l_discount >= Decimal128(Some(5),15,2), lineitem.l_discount <= Decimal128(Some(7),15,2), lineitem.l_quantity < Decimal128(Some(2400),15,2)]
+04)------Filter: lineitem.l_shipdate >= Date32("1994-01-01") AND lineitem.l_shipdate < Date32("1995-01-01") AND CAST(lineitem.l_discount AS Float64) >= Float64(0.049999999999999996) AND CAST(lineitem.l_discount AS Float64) <= Float64(0.06999999999999999) AND lineitem.l_quantity < Decimal128(Some(2400),15,2)
+05)--------TableScan: lineitem projection=[l_quantity, l_extendedprice, l_discount, l_shipdate], partial_filters=[lineitem.l_shipdate >= Date32("1994-01-01"), lineitem.l_shipdate < Date32("1995-01-01"), CAST(lineitem.l_discount AS Float64) >= Float64(0.049999999999999996), CAST(lineitem.l_discount AS Float64) <= Float64(0.06999999999999999), lineitem.l_quantity < Decimal128(Some(2400),15,2)]
 physical_plan
 01)ProjectionExec: expr=[sum(lineitem.l_extendedprice * lineitem.l_discount)@0 as revenue]
 02)--AggregateExec: mode=Final, gby=[], aggr=[sum(lineitem.l_extendedprice * lineitem.l_discount)]
 03)----CoalescePartitionsExec
 04)------AggregateExec: mode=Partial, gby=[], aggr=[sum(lineitem.l_extendedprice * lineitem.l_discount)]
 05)--------CoalesceBatchesExec: target_batch_size=8192
-06)----------FilterExec: l_shipdate@3 >= 1994-01-01 AND l_shipdate@3 < 1995-01-01 AND l_discount@2 >= Some(5),15,2 AND l_discount@2 <= Some(7),15,2 AND l_quantity@0 < Some(2400),15,2, projection=[l_extendedprice@1, l_discount@2]
+06)----------FilterExec: l_shipdate@3 >= 1994-01-01 AND l_shipdate@3 < 1995-01-01 AND CAST(l_discount@2 AS Float64) >= 0.049999999999999996 AND CAST(l_discount@2 AS Float64) <= 0.06999999999999999 AND l_quantity@0 < Some(2400),15,2, projection=[l_extendedprice@1, l_discount@2]
```

Comment:
This is incorrect. `l_decimal` is a decimal and the literals in the query are also decimals. There should not be a case from an exact type to an approximate type here.

```sql
and l_discount between 0.05 and 0.07
```

Reply:
@andygrove I agree that 0.05  should be decimal, but it's a float in DataFusion, unless `parse_float_as_decimal` flag is enabled.
Are you implying that you expect `parse_float_as_decimal` to be true by default?

Reply:
Yes, that should be enabled by default now. I vaguely remember this being enabled a long time ago. I will try and look today.

Reply:
Yes, that should be enabled by default now. I vaguely remember this being enabled a long time ago. I will try and look today.

Reply:
I suspect that an analysis or optimizer rule (perhaps type coercion and/or constant folding) is casting the literal floats in the SQL to decimal because the expression is comparing `l_decimal` (a decimal) to this value.

Reply:
There is no special handling for that. It worked for this query because of the incorrect coercion which this PR is fixing. So affecting this query should be _expected_, and is not a bug.

Reply:
q6 should be comparing decimals in the table and decimal literals in the query. I agree that the way that DataFusion is currently implementing this is not ideal, but the query result is correct. This PR currently makes the query return an incorrect result because it is casting the decimals to an approximate floating point and then comparing those values to an approximate floating point representation of the decimal literals in the query.

Reply:
By focusing on q6 which seems to work by combination of two bugs, we conveniently avoid debating the merit  of the change (or lack thereof).

Reply:
I filed https://github.com/apache/datafusion/issues/14612#issuecomment-2651278826 for the SQL parsing issue and linked to the type coercion issue.